### ISSUE TITLE
Don't warn when `Optional.of()` is called with a `@Nullable` value

### ISSUE THREAD

**Issue by donalmurtagh** [uber/NullAway#1273]
if the return type of `something()` is `@Nullable` then the following will generate the familiar warning

> passing @Nullable parameter where @NonNull is required

```java
var maybeNull = something();
var notEmpty = Optional.of(maybeNull);  // generates a nullable warning
```

In my case `something()` is a library method that returns a `@Nullable`, so the only way I know of to fix the warning
is to add an appropriate null check before calling `Optional.of`, e.g.

```java
var notNull = Objects.requireNonNull(something());
var notEmpty = Optional.of(notNull);
```

However, the implementation of `Optional.of` calls `Object.requireNonNull` on the argument, so calling it before `Optional.of` is redundant.

1. Should NullAway disable nullable warnings for `Optional.of` by default?
2. Is there a configuration option that allows me to disable nullable warnings for all calls to `Optional.of` in my project?

**Comment by msridhar** [comment]
Hi @donalmurtagh I don't think it'd be a good idea to disable this check, since `Optional.of` does require a `@NonNull` argument and is guaranteed to fail if you pass `null`.  I'm curious why you would want to disable this across your project.  Do you have lots of cases where you are calling `Optional.of` with some `@Nullable` argument, but somehow you know that argument cannot be null?

**Comment by msridhar** [comment]
Which `libraryMethod` is this?  There is potentially a way they could annotate that method using `@Contract` to eliminate this warning entirely.

In the meantime, I think your other option is to add a [downcast method](https://github.com/uber/NullAway/wiki/Suppressing-Warnings#downcasting) and use that.  It's best practice to still have a null check in the downcast method and do _something_ if the arg is null, but strictly speaking you don't have to do that.

FWIW, from a performance standpoint, my guess is the extra null check would cost basically nothing in hot code; the JIT should do enough inlining to detect the redundancy (in fact, `requireNonNull` is [annotated as `@ForceInline`](https://github.com/openjdk/jdk/blob/c935d1ce1c42ce98cc6ceffaa4f47eb2dba24dfd/src/java.base/share/classes/java/util/Objects.java#L217) which likely means it will always be inlined when running on OpenJDK).

**Comment by scordio** [comment]
I think the existing [`@Contract`](https://docs.spring.io/spring-framework/docs/current-SNAPSHOT/javadoc-api/org/springframework/lang/Contract.html) annotation might help, so I played with it in this branch: https://github.com/scordio/spring-security/commit/15ae3999e87b63b2528b8767bceda2005e3ea9d6

@donalmurtagh could you maybe build my branch locally and see if it solves your use case? If yes, I'm happy to raise a PR.

You'd need to set this flag in your NullAway config:

```
-XepOpt:NullAway:CustomContractAnnotations=org.springframework.lang.Contract
```

**Comment by donalmurtagh** [comment]
@scordio I upgraded to Spring Security 7.1.0 today and the problem no longer exists thanks to the `@Contract` annotation. Much appreciated.
